In [1]:
import pandas as pd

url = "https://www.sharkattackfile.net/spreadsheets/GSAF5.xls"

df_shark_attacks = pd.read_excel(url)
df_shark_attacks


Date    Year        Type             Country  \
0      10th January  2026.0  Unprovoked           Australia   
1       8th January  2026.0  Unprovoked   US Virgin Islands   
2      3rd January   2026.0  Unprovoked       New Caledonia   
3     21st December  2025.0  Unprovoked                 USA   
4     12th December  2025.0  Unprovoked                 USA   
...             ...     ...         ...                 ...   
7060    Before 1903     0.0  Unprovoked           AUSTRALIA   
7061    Before 1903     0.0  Unprovoked           AUSTRALIA   
7062      1900-1905     0.0  Unprovoked                 USA   
7063      1883-1889     0.0  Unprovoked              PANAMA   
7064      1845-1853     0.0  Unprovoked  CEYLON (SRI LANKA)   

                             State                             Location  \
0                              NSW                         Avalon Beach   
1     Fredricksted Island St Croix                         Dorsch Beach   
2                             Kélé          Between Bourail and Moindou   
3                       California           Lovers Point Pacific Grove   
4         Sonoma County California                         Salmon Creek   
...                            ...                                  ...   
7060             Western Australia                          Roebuck Bay   
7061             Western Australia                                  NaN   
7062                North Carolina                       Ocracoke Inlet   
7063                           NaN                 Panama Bay 8ºN, 79ºW   
7064              Eastern Province  Below the English fort, Trincomalee   

          Activity                   Name Sex  Age  ...  \
0          Surfing           Paul Stanton   M    ?  ...   
1       Snorkeling          Arlene Lillis   F   56  ...   
2     Scuba Diving                Unknown   M    ?  ...   
3         Swimming              Erica Fox   F   55  ...   
4          Surfing                Unknown   M    ?  ...   
...            ...                    ...  ..  ...  ...   
7060        Diving                   male   M  NaN  ...   
7061  Pearl diving                  Ahmun   M  NaN  ...   
7062      Swimming  Coast Guard personnel   M  NaN  ...   
7063           NaN        Jules Patterson   M  NaN  ...   
7064      Swimming                   male   M   15  ...   

                         Species   \
0                         Unknown   
1                        Unknown    
2                         Unknown   
3               Great White Shark   
4     Suspected Great White Shark   
...                           ...   
7060                          NaN   
7061                          NaN   
7062                          NaN   
7063                          NaN   
7064                          NaN   

                                                 Source  \
0                                        Bob Myatt GSAF   
1          Todd Smith: KevinMcMurray Trackingsharks.com   
2                            Andy Currie: Province Sud:   
3     Kevin McMurray Tracking sharks.com: Ralph Coll...   
4     Kevin McMurray Tracking sharks.com:Andrew Curr...   
...                                                 ...   
7060                   H. Taunton; N. Bartlett,  p. 234   
7061              H. Taunton; N. Bartlett,  pp. 233-234   
7062               F. Schwartz, p.23; C. Creswell, GSAF   
7063                                The Sun, 10/20/1938   
7064                                         S.W. Baker   

                                 pdf  \
0                                NaN   
1                                NaN   
2                                NaN   
3                                NaN   
4                                NaN   
...                              ...   
7060          ND-0005-RoebuckBay.pdf   
7061               ND-0004-Ahmun.pdf   
7062  ND-0003-Ocracoke_1900-1905.pdf   
7063      ND-0002-JulesPatterson.pdf   
7064              ND-0001-Ceylon.pdf   

             

In [2]:
df_shark_attacks.shape


(7065, 23)

In [3]:
print(df_shark_attacks.isnull().sum())

Date                 0
Year                 2
Type                18
Country             50
State              487
Location           567
Activity           585
Name               219
Sex                579
Age               2995
Injury              35
Fatal Y/N          561
Time              3527
Species           3131
Source              20
pdf                266
href formula       271
href               269
Case Number        267
Case Number.1      268
original order     266
Unnamed: 21       7064
Unnamed: 22       7063
dtype: int64


Comprobacion de nulo de cada columna

Gestion de columnas y duplicados

In [4]:
df_shark_attacks.columns = [x .lower().strip() for x in df_shark_attacks.columns]
df_shark_attacks.columns

Index(['date', 'year', 'type', 'country', 'state', 'location', 'activity',
       'name', 'sex', 'age', 'injury', 'fatal y/n', 'time', 'species',
       'source', 'pdf', 'href formula', 'href', 'case number', 'case number.1',
       'original order', 'unnamed: 21', 'unnamed: 22'],
      dtype='object')

Convertimos en minusculas las columnas

In [5]:
df_shark_attacks.rename(columns={'species ': 'species', 'sex ': 'sex'}, inplace=True)

Modificacion de los nombres, ya que tenian espacios

In [6]:
columnas_a_borrar = ['unnamed: 21', 'pdf', 'href','href formula','case number','case number.1','original order', 'unnamed: 21', 'unnamed: 22']
df_shark_attacks.drop(columns=[col for col in columnas_a_borrar if col in df_shark_attacks.columns], inplace=True)

Borrado de columnas no necesarias para el analisis

In [7]:
df_shark_attacks.drop_duplicates(inplace=True) 
df_shark_attacks.reset_index(drop=True,inplace=True)

Borrado de duplicados y reseteo de indice

In [8]:
df_shark_attacks['year'] = df_shark_attacks['year'].fillna(0)
df_shark_attacks['year'] = pd.to_numeric(df_shark_attacks['year'], errors='coerce').fillna(0).astype(int)

Imputacion de los nulos a 0 y convertir la columna 'year' a entero

In [9]:
print( df_shark_attacks['year'].min())
print( df_shark_attacks['year'].max())

0
2026


Comprobacion de minimo y maximo en columna year

In [10]:
year_actual = df_shark_attacks[df_shark_attacks['year'] > 1900]
year_actual

date  year        type            country  \
0             10th January  2026  Unprovoked          Australia   
1              8th January  2026  Unprovoked  US Virgin Islands   
2             3rd January   2026  Unprovoked      New Caledonia   
3            21st December  2025  Unprovoked                USA   
4            12th December  2025  Unprovoked                USA   
...                    ...   ...         ...                ...   
6296           17-Jul-1901  1901     Invalid              ITALY   
6297  Reported 29-Jun-1901  1901     Invalid             YEMEN    
6298           24-Jun-1901  1901  Unprovoked        PHILIPPINES   
6299           30-Jan-1901  1901  Unprovoked          AUSTRALIA   
6300           Summer 1901  1901  Unprovoked                USA   

                             state                                location  \
0                              NSW                            Avalon Beach   
1     Fredricksted Island St Croix                            Dorsch Beach   
2                             Kélé             Between Bourail and Moindou   
3                       California              Lovers Point Pacific Grove   
4         Sonoma County California                            Salmon Creek   
...                            ...                                     ...   
6296                      Syracuse                        Capo Santa Croce   
6297                          Aden                                     NaN   
6298              Western Viscayas                 Island of Panay, Iliolo   
6299                    Queensland                                Brisbane   
6300                       Florida  Indian River Inlet / Fort Pierce Inlet   

                          activity  \
0                          Surfing   
1                       Snorkeling   
2                     Scuba Diving   
3                         Swimming   
4                          Surfing   
...                            ...   
6296                      Swimming   
6297  Diving around anchored liner   
6298                      Swimming   
6299                       Bathing   
6300                       Bathing   

                                                   name sex  age  \
0                                          Paul Stanton   M    ?   
1                                         Arlene Lillis   F   56   
2                                               Unknown   M    ?   
3                                             Erica Fox   F   55   
4                                               Unknown   M    ?   
...                                                 ...  ..  ...   
6296                                  Antonio Tornatori   M   20   
6297                                                boy   M  NaN   
6298  S. McKie, apprentice 1st class on the U.S. Nav...   M  NaN   
6299                                      John Thompson   M  NaN   
6300                                    Michael O'Brien   M  NaN   

                                                 injury fatal y/n     time  \
0                           puncture mark to left thumb         N  0540hrs   
1      Left arm torn off in the attack below the elbow          Y  1628hrs   
2                               Injuries to upper limbs         N        ?   
3     Taken by shark body recovered with multiple in...         Y  1200hrs   
4                                           Hand Injury         N  0800hrs   
...                                                 ...       ...      ...   
6296     Disappeared, but shark involvement unconfirmed       NaN      NaN   
6297                                                NaN       NaN      NaN   
6298    Left thigh stripped of flesh 4" above the knee          N    15h00   
6299                                Thigh bitten, FATAL         Y      NaN   
6300                                     Abdomen bitten         N      NaN   

                          species  \
0                         Unknow

Creacion de variable para utilizar los años desde 1900

In [11]:
df_shark_attacks['year'] = pd.to_numeric(df_shark_attacks['year'], errors='coerce')

convertimos year a numerico

In [12]:
for col in ['country','location']:
    df_shark_attacks[col] = df_shark_attacks[col].fillna('UNKNOWN').astype(str).str.strip()

df_shark_attacks['country'] = df_shark_attacks['country'].str.upper()
df_shark_attacks['location'] = df_shark_attacks['location'].str.title()

Limpieza de strings a mayuscula en country y añadir titulos. Tambien he rellenado los nulos con UNKNOWN

In [13]:
df_shark_attacks['name'] = df_shark_attacks['name'].fillna('Anonymous').astype(str).str.strip()
df_shark_attacks.loc[df_shark_attacks['name'].str.lower().isin(['male', 'female']), 'name'] = 'Anonymous'

Rellenar los nulos de la columna name con Anonymous
Reemplazar valores male y female como Anonymous


In [14]:
df_shark_attacks['age'].value_counts()

age
16          92
19          90
17          89
18          85
17          85
            ..
10 or 12     1
31 or 33     1
2½           1
1            1
13 or 14     1
Name: count, Length: 251, dtype: int64

comprobar valores en columna age para saber como actuar

In [15]:
df_shark_attacks['age'] = df_shark_attacks['age'].astype(str)

Cambiar la columna age que  tipo string ya que tiene muchos valores como nan

In [16]:
df_shark_attacks['age_clean'] = pd.to_numeric(df_shark_attacks['age'], errors='coerce')

Creacion de una columna nueva de Age = age_clean

In [17]:
df_shark_attacks['age_clean'].isnull().sum()

np.int64(3145)

Comprobacion de nulos por imputar

In [18]:
age_mediana = df_shark_attacks['age_clean'].median()
age_mediana

24.0

Calculamos la mediana de la edad para sacar conclusiones 

In [19]:
df_shark_attacks['age_clean'] = df_shark_attacks['age_clean'].fillna(age_mediana)

Rellenamos los NaN con la mediana

In [23]:
df_shark_attacks['age_clean'] = df_shark_attacks['age_clean'].astype(int)

convertimos a entero para quitar los decimales

In [24]:
df_shark_attacks['age_clean'].median()

24.0

Comprobacion de datos en Age_clean

In [25]:
df_shark_attacks.drop(columns=['age'], inplace=True)

Borrado de la tabla age antigua

In [26]:
df_shark_attacks

date  year        type             country  \
0      10th January  2026  Unprovoked           AUSTRALIA   
1       8th January  2026  Unprovoked   US VIRGIN ISLANDS   
2      3rd January   2026  Unprovoked       NEW CALEDONIA   
3     21st December  2025  Unprovoked                 USA   
4     12th December  2025  Unprovoked                 USA   
...             ...   ...         ...                 ...   
7059    Before 1903     0  Unprovoked           AUSTRALIA   
7060    Before 1903     0  Unprovoked           AUSTRALIA   
7061      1900-1905     0  Unprovoked                 USA   
7062      1883-1889     0  Unprovoked              PANAMA   
7063      1845-1853     0  Unprovoked  CEYLON (SRI LANKA)   

                             state                             location  \
0                              NSW                         Avalon Beach   
1     Fredricksted Island St Croix                         Dorsch Beach   
2                             Kélé          Between Bourail And Moindou   
3                       California           Lovers Point Pacific Grove   
4         Sonoma County California                         Salmon Creek   
...                            ...                                  ...   
7059             Western Australia                          Roebuck Bay   
7060             Western Australia                              Unknown   
7061                North Carolina                       Ocracoke Inlet   
7062                           NaN                 Panama Bay 8ºn, 79ºw   
7063              Eastern Province  Below The English Fort, Trincomalee   

          activity                   name sex  \
0          Surfing           Paul Stanton   M   
1       Snorkeling          Arlene Lillis   F   
2     Scuba Diving                Unknown   M   
3         Swimming              Erica Fox   F   
4          Surfing                Unknown   M   
...            ...                    ...  ..   
7059        Diving              Anonymous   M   
7060  Pearl diving                  Ahmun   M   
7061      Swimming  Coast Guard personnel   M   
7062           NaN        Jules Patterson   M   
7063      Swimming              Anonymous   M   

                                                 injury fatal y/n     time  \
0                           puncture mark to left thumb         N  0540hrs   
1      Left arm torn off in the attack below the elbow          Y  1628hrs   
2                               Injuries to upper limbs         N        ?   
3     Taken by shark body recovered with multiple in...         Y  1200hrs   
4                                           Hand Injury         N  0800hrs   
...                                                 ...       ...      ...   
7059                                              FATAL         Y      NaN   
7060                                              FATAL         Y      NaN   
7061                                              FATAL         Y      NaN   
7062                                              FATAL         Y      NaN   
7063  FATAL. "Shark bit him in half, carrying away t...         Y      NaN   

                          species  \
0                         Unknown   
1                        Unknown    
2                         Unknown   
3               Great White Shark   
4     Suspected Great White Shark   
...                           ...   
7059                          NaN   
7060                          NaN   
7061                          NaN   
7062                          NaN   
7063                          NaN   

                                                 source  age_clean  
0                                        Bob Myatt GSAF         24  
1          Todd Smith: KevinMcMurray Trackingsharks.com         56  
2                            Andy Currie: Province Sud:         24  
3     Kevin McMurray Tracking sharks.com: Ralph Coll...         55  
4     Kevin McMurray Tracking sharks.com:Andrew Curr...         2

In [27]:
df_shark_attacks.rename(columns={'fatal y/n':'fatal'}, inplace=True)

Cambiar nombre a la columna

In [28]:
df_shark_attacks['fatal'] = df_shark_attacks['fatal'].astype(str).str.strip().str.upper()


Converitmos a mayusculas todas las filas de la columna Fatal

In [29]:
df_shark_attacks['fatal'].unique()

array(['N', 'Y', 'F', 'M', 'NAN', 'NQ', 'UNKNOWN', '2017', 'Y X 2'],
      dtype=object)

Comprobamos los valores unicos en la columna fatal

In [30]:
#Funcion para definir que solo queremos valores 'Y' o 'N'

def clean_fatal(valor):
    if valor == 'Y':
        return 'Y'
    elif valor == 'N':
        return 'N'
    else:
        return 'UNKNONW'


In [31]:
df_shark_attacks['fatal'] = df_shark_attacks['fatal'].apply(clean_fatal)
df_shark_attacks['fatal'].value_counts()


fatal
N          4934
Y          1487
UNKNONW     643
Name: count, dtype: int64

Comprobacion de valores en columna 'fatal'

In [32]:
df_shark_attacks['species'].value_counts()

species
White shark                                           194
Shark involvement prior to death was not confirmed    105
Invalid                                               102
Shark involvement not confirmed                        92
Tiger shark                                            89
                                                     ... 
234-lb shark                                            1
Said to involve a 6 m to 7.3 m [20' to 24'] shark       1
Said to be a 7.6 m [25'] shark                          1
Thought to involve a porbeagle or mako shark            1
Not stgated                                             1
Name: count, Length: 1736, dtype: int64

Comprobamos como estan escritos las principales nombres de la especies

In [33]:
df_shark_attacks['species'] = df_shark_attacks['species'].fillna('Unknown').astype(str)

Rellenamos los nulos con Unknown y a tipo str

In [34]:
#Funcion para mejorar los valores dentro de la tabla species

def categoria_species(valor):
    valor = valor.lower()  #convertimos todo a minuscula

    if 'white' in valor:
        return 'White Shark'
    elif 'tiger' in valor:
        return 'Tiger Shark'
    elif 'bull' in valor:
        return 'Bull Shark'
    elif 'hammerhead' in valor:
        return 'Hammerhead Shark'
    elif 'shark' in valor:
        return 'Other Shark'
    else:
        
        return 'Unknown'
    
df_shark_attacks['species'] = df_shark_attacks['species'].apply(categoria_species)

He creado solo 5 categorias de tiburones para poder sacar medias de cada especie y Unknown para los que no se puede saber la especie o no esta bien descrito

In [35]:
print(df_shark_attacks['species'].value_counts())

species
Unknown             3431
Other Shark         2250
White Shark          756
Tiger Shark          344
Bull Shark           234
Hammerhead Shark      49
Name: count, dtype: int64


Comprobacion de la tabla species

In [36]:
df_shark_attacks

date  year        type             country  \
0      10th January  2026  Unprovoked           AUSTRALIA   
1       8th January  2026  Unprovoked   US VIRGIN ISLANDS   
2      3rd January   2026  Unprovoked       NEW CALEDONIA   
3     21st December  2025  Unprovoked                 USA   
4     12th December  2025  Unprovoked                 USA   
...             ...   ...         ...                 ...   
7059    Before 1903     0  Unprovoked           AUSTRALIA   
7060    Before 1903     0  Unprovoked           AUSTRALIA   
7061      1900-1905     0  Unprovoked                 USA   
7062      1883-1889     0  Unprovoked              PANAMA   
7063      1845-1853     0  Unprovoked  CEYLON (SRI LANKA)   

                             state                             location  \
0                              NSW                         Avalon Beach   
1     Fredricksted Island St Croix                         Dorsch Beach   
2                             Kélé          Between Bourail And Moindou   
3                       California           Lovers Point Pacific Grove   
4         Sonoma County California                         Salmon Creek   
...                            ...                                  ...   
7059             Western Australia                          Roebuck Bay   
7060             Western Australia                              Unknown   
7061                North Carolina                       Ocracoke Inlet   
7062                           NaN                 Panama Bay 8ºn, 79ºw   
7063              Eastern Province  Below The English Fort, Trincomalee   

          activity                   name sex  \
0          Surfing           Paul Stanton   M   
1       Snorkeling          Arlene Lillis   F   
2     Scuba Diving                Unknown   M   
3         Swimming              Erica Fox   F   
4          Surfing                Unknown   M   
...            ...                    ...  ..   
7059        Diving              Anonymous   M   
7060  Pearl diving                  Ahmun   M   
7061      Swimming  Coast Guard personnel   M   
7062           NaN        Jules Patterson   M   
7063      Swimming              Anonymous   M   

                                                 injury fatal     time  \
0                           puncture mark to left thumb     N  0540hrs   
1      Left arm torn off in the attack below the elbow      Y  1628hrs   
2                               Injuries to upper limbs     N        ?   
3     Taken by shark body recovered with multiple in...     Y  1200hrs   
4                                           Hand Injury     N  0800hrs   
...                                                 ...   ...      ...   
7059                                              FATAL     Y      NaN   
7060                                              FATAL     Y      NaN   
7061                                              FATAL     Y      NaN   
7062                                              FATAL     Y      NaN   
7063  FATAL. "Shark bit him in half, carrying away t...     Y      NaN   

          species                                             source  \
0         Unknown                                     Bob Myatt GSAF   
1         Unknown       Todd Smith: KevinMcMurray Trackingsharks.com   
2         Unknown                         Andy Currie: Province Sud:   
3     White Shark  Kevin McMurray Tracking sharks.com: Ralph Coll...   
4     White Shark  Kevin McMurray Tracking sharks.com:Andrew Curr...   
...           ...                                                ...   
7059      Unknown                   H. Taunton; N. Bartlett,  p. 234   
7060      Unknown              H. Taunton; N. Bartlett,  pp. 233-234   
7061      Unknown               F. Schwartz, p.23; C. Creswell, GSAF   
7062      Unknown                                The Sun, 10/20/1938   
7063      Unknown                                         S.W. Baker   

      age_clean  
0            24  
1 

RESUMEN DEL TRABAJO DE LIMPIEZA Y ESTANDARIZACION DE DATOS

- Borrado de columnas no necesarias para el analisis
- Borrado de duplicados y reseteo de indice
- Imputacion de los nulos a 0 y convertir la columna 'year' a entero
- Tengo que limpiar mejor la columna year ********
- Limpieza de strings a mayuscula en country y añadir titulos. Tambien he rellenado los nulos con UNKNOWN
- Rellenar los nulos de la columna name con Anonymous
- Reemplazar valores male y female como Anonymous
- Cambiar la columna age a tipo string ya que tiene muchos valores como nan
- Creacion de una columna nueva de Age = age_clean
- Convertimos a entero para quitar los decimales
- Calculamos la mediana de la edad para sacar conclusiones 
- Borrado de la tabla age antigua
- Cambiar nombre a la columna Fatal Y/N por Fatal
- Converitmos a mayusculas todas las filas de la columna Fatal
- Comprobamos los valores unicos en la columna fatal
- Comprobamos como estan escritos las principales nombres de la especies
- Rellenamos los nulos con Unknown y a tipo str
- He creado solo 5 categorias de tiburones con una funcion para poder sacar mejores conclusiones y Unknown para los que no se puede saber la especie o no esta bien descrito

CONCLUSIONES SEGUN LOS DATOS QUE HE LIMPIADO

Conclusion por letalidad y especie: 
 Tras limpiar y estandarizar la columna Fatal, los datos sugieren que la tasa de supervivencia es significativamente alta con 4394 casos como N. (Podemos sacar porcentes exacto)

Respecto a las especies, aunque el tiburón blanco aparece frecuentemente en los registros, un gran porcentaje de ataques provienen de especies no identificadas (Unknown), lo que sugiere que ciertas especies pueden estar sesgada por la facilidad de identificación visual, mientras que la letalidad real no siempre correlaciona con la frecuencia de ataques. (Podemos sacar los porcentages)

Conclusion Demografica:
Se han sufrido ataques en muchos rangos de edad pero al procesar la columna age_clean y convertirla a valores numéricos enteros (inputando los nulos con la mediana), se observa una concentración de datos en el rango de adultos jóvenes y adolescentes. (Comparar con activity para ver si esta correlacionado)

Concluion Historica:

He de mejorar primero la columna year para sacar conclusiones

In [37]:
df_usa = df_shark_attacks[df_shark_attacks['country'] == 'USA'] 

He filtadro la informacion de USA

In [39]:
activities = df_usa['activity'].astype(str)
activities.value_counts()

activity
Surfing                                                                                                     664
Swimming                                                                                                    370
nan                                                                                                         155
Fishing                                                                                                     140
Wading                                                                                                      124
                                                                                                           ... 
Fishing for basking sharks                                                                                    1
Shark tagging                                                                                                 1
Kitesurfing                                                                                    

Comprobacion de accidentes mientras se hace alguna actividad en USA

In [40]:
condicion_1 = activities.str.contains('diving', case=False, regex=False)
condicion_2 = activities.str.contains('scuba', case=False, regex=False)
condicion_3 = activities.str.contains('snorkel', case=False, regex=False)
df_usa_diving = df_usa[condicion_1 | condicion_2 | condicion_3]

df_usa_diving['activity'].value_counts().head()

activity
Snorkeling                 41
Diving                     33
Scuba diving               24
Free diving                11
Free diving for abalone     6
Name: count, dtype: int64

conteo de ataques de tiburon con las actividades usaremos para nuestra empresa

In [41]:
condicion_4 = activities.str.contains('surf', case=False, regex=False)
condicion_5 = activities.str.contains('swimming', case=False, regex=False)
condicion_6= activities.str.contains('fishing', case=False, regex=False)
df_usa_others_activities = df_usa[condicion_4 | condicion_5 | condicion_6]
df_usa_others_activities['activity'].value_counts().head()

activity
Surfing         664
Swimming        370
Fishing         140
Spearfishing     64
Body surfing     24
Name: count, dtype: int64

Podemos comparar el resto de actividades para demostrar que es mas seguro hacer buceo que las que se muestran en la tabla

In [42]:

meses = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun','Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec']

def buscar_mes(date):
   
    date = str(date).title()
    
    
    for mes in meses:
        
        
        if mes in date: 
            return mes 
            
    return None 


df_usa_diving['mes'] = df_usa_diving['date'].apply(buscar_mes)


print(df_usa_diving['mes'].unique())


['Sep' 'Jul' None 'Dec' 'Oct' 'Aug' 'Jan' 'Nov' 'Mar' 'Apr' 'May' 'Jun'
 'Feb']


C:\Users\manue\AppData\Local\Temp\ipykernel_28368\435822726.py:17: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_usa_diving['mes'] = df_usa_diving['date'].apply(buscar_mes)


Funcion para filtrar meses

In [43]:
df_usa_diving['mes'] = df_usa_diving['date'].apply(buscar_mes)


df_usa_diving['mes'].value_counts()

C:\Users\manue\AppData\Local\Temp\ipykernel_28368\1129378743.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_usa_diving['mes'] = df_usa_diving['date'].apply(buscar_mes)


mes
Sep    31
Aug    30
Jul    27
Oct    21
Jun    17
May    14
Nov    14
Mar    11
Dec     9
Feb     9
Jan     8
Apr     6
Name: count, dtype: int64

He filtrado por meses, para saber cuando hay mas ataques de tiburones y a si saber cuando es mas probable ver alguno

In [44]:
fatality_state = df_usa_diving.groupby(['state', 'injury','mes'], as_index=False).size()
fatality_state['state'].value_counts()

state
Florida              65
California           59
Hawaii               47
North Carolina        6
US Virgin Islands     2
Connecticut           1
Guam                  1
Delaware              1
Maryland              1
Louisiana             1
Massachusetts         1
New Jersey            1
New Mexico            1
Maine                 1
North Carolina        1
Palmyra Atoll         1
Rhode Island          1
Puerto Rico           1
Samoa                 1
South Carolina        1
Texas                 1
Virgin Islands        1
Name: count, dtype: int64

He agrupado por state y injury y mes para saber donde hay mas ataques de tiburones

In [45]:
top_states =['Florida', 'California', 'Hawaii']

He seleccionado Florida, California y Hawaii, ya que aparecen con diferencia los estados con mas ataques

In [46]:
df_top3 = df_usa_diving[df_usa_diving['state'].isin(top_states)]


- He definido el top 3

In [47]:
ranking = df_top3.groupby(['state', 'mes','species'], as_index=False).size()
ranking

state  mes      species  size
0   California  Apr  White Shark     1
1   California  Aug      Unknown     2
2   California  Aug  White Shark     5
3   California  Dec  Other Shark     1
4   California  Dec  White Shark     4
..         ...  ...          ...   ...
65      Hawaii  Oct  Tiger Shark     2
66      Hawaii  Oct      Unknown     3
67      Hawaii  Oct  White Shark     2
68      Hawaii  Sep  Other Shark     2
69      Hawaii  Sep      Unknown     2

[70 rows x 4 columns]

He agrupado por mes y estado y especie para saber en que epoca se producen mas ataques y de que tipo de tiburones
podemos obserbar que en California los meses con mas ataques son septimbre con 14, 13 de ellos tiburones blancos, julio con 8 y 7 de ellos tiburones blancos, y en agosto 7 y 5 de ellos tiburones blancos
En florida en septiembre  con 14 y 13 de ellos blancos. no hay mas de tiburones blancos
En hawaii solo 2 de tiburones blancos en octubre

Podemos decir que hay una gran variedad de especies en estos 3 estados


In [48]:
df_shark_attacks['injury']

0                             puncture mark to left thumb
1        Left arm torn off in the attack below the elbow 
2                                 Injuries to upper limbs
3       Taken by shark body recovered with multiple in...
4                                             Hand Injury
                              ...                        
7059                                                FATAL
7060                                                FATAL
7061                                                FATAL
7062                                                FATAL
7063    FATAL. "Shark bit him in half, carrying away t...
Name: injury, Length: 7064, dtype: object

In [49]:
def riesgo(risk):
    risk = str(risk).lower()
    
   
    if 'fatal' in risk:
        return 'FATAL'
    

    elif 'no injury' in risk or 'not injured' in risk:
        return 'SIN HERIDAS'
    
    
    elif 'lacerat' in risk or 'bit' in risk or 'severed' in risk or 'injur' in risk:
        return 'HERIDA FISICA'
        
    
    else:
        return 'OTRO/LEVE'


df_seguridad = df_top3 
df_seguridad['Tipo_Riesgo'] = df_seguridad['injury'].apply(riesgo)


C:\Users\manue\AppData\Local\Temp\ipykernel_28368\762037663.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_seguridad['Tipo_Riesgo'] = df_seguridad['injury'].apply(riesgo)


He hecho una funcion para filtrar por tipo de heridas

In [50]:
df_seguridad['Tipo_Riesgo']

19      HERIDA FISICA
30      HERIDA FISICA
40          OTRO/LEVE
65      HERIDA FISICA
89      HERIDA FISICA
            ...      
6779    HERIDA FISICA
6954    HERIDA FISICA
6988    HERIDA FISICA
6992    HERIDA FISICA
6994      SIN HERIDAS
Name: Tipo_Riesgo, Length: 187, dtype: object

In [51]:
df_seguridad = df_top3


df_seguridad['Tipo_Riesgo'] = df_seguridad['injury'].apply(riesgo)


grupo_seguridad = df_seguridad.groupby(['state', 'Tipo_Riesgo'], as_index=False).size()


grupo_seguridad.rename(columns={'size': 'Total_Casos'}, inplace=True)


grupo_seguridad = grupo_seguridad.sort_values(by=['state', 'Total_Casos'], ascending=[True, False])


print(grupo_seguridad)

         state    Tipo_Riesgo  Total_Casos
1   California  HERIDA FISICA           45
3   California    SIN HERIDAS           10
2   California      OTRO/LEVE            5
0   California          FATAL            3
5      Florida  HERIDA FISICA           41
6      Florida      OTRO/LEVE           12
7      Florida    SIN HERIDAS           11
4      Florida          FATAL            8
9       Hawaii  HERIDA FISICA           33
8       Hawaii          FATAL            8
11      Hawaii    SIN HERIDAS            6
10      Hawaii      OTRO/LEVE            5


C:\Users\manue\AppData\Local\Temp\ipykernel_28368\2072395966.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_seguridad['Tipo_Riesgo'] = df_seguridad['injury'].apply(riesgo)


He agrupado por estado y tipo de riesgo, para saber que medidas de seguridad aplicar y tambien donde hay mas posibilidades de sufrir algun accidente dependiendo del estado

Podemos ver que donde mas heridas fisicas hay es en California con 45, pero tambien donde menor fatalidad hay. En floridad tenemos 41 heridas fisicas en cambio hay mas posibilades de fatalidad
En Hawaii podemos observar que es donde menos heridas fisicas hay y hay una fatalidad igual a la de Florida. Hay que tener en cuenta que en Hawaii hay menos avistamientos de tiburones pero una fatalidad mas alta

# Resumen del trabajo del dia 21/1/1
  - Limpieza de datos
      - He filtadro la informacion de USA
      - Comprobacion de accidentes mientras se hace alguna actividad en USA 
      - Conteo de ataques de tiburon con las actividades usaremos para nuestra empresa
      - Comparacion de el resto de actividades para demostrar que es mas seguro hacer buceo que las que se muestran en la tabla
      - He creado una funcion para filtrar por meses
      - He filtrado por meses, para saber cuando hay mas ataques de tiburones y a si saber cuando es mas probable ver alguno
      - He agrupado por state y injury y mes para saber donde hay mas ataques de tiburones para saber donde hay mas tiburones y cuando es mas probable verlos
      - He seleccionado Florida, California y Hawaii, ya que aparecen con diferencia los estados con mas ataques y lo he definido

  - Agrupamientos 
      - He agrupado por mes y estado y especie para saber en que epoca se producen mas ataques y de que tipo de tiburones
        podemos obserbar que en California los meses con mas ataques son septimbre con 14, 13 de ellos tiburones blancos, julio con 8 y 7 de ellos tiburones blancos, y en agosto 7 y 5 de ellos tiburones blancos
        En florida en septiembre  con 14 y 13 de ellos blancos. no hay mas de tiburones blancos
        En hawaii solo 2 de tiburones blancos en octubre

        Podemos decir que hay una gran variedad de especies en estos 3 estados
      
      - He hecho una funcion para filtrar por tipo de heridas
      - He agrupado por estado y tipo de riesgo, para saber que medidas de seguridad aplicar y tambien donde hay mas posibilidades de sufrir algun accidente dependiendo del estado

      - Podemos ver que donde mas heridas fisicas hay es en California con 45, pero tambien donde menor fatalidad hay. En floridad tenemos 41 heridas fisicas en cambio hay mas posibilades de fatalidad
       En Hawaii podemos observar que es donde menos heridas fisicas hay y hay una fatalidad igual a la de Florida. Hay que tener en cuenta que en Hawaii hay menos avistamientos de tiburones pero una fatalidad mas alta

# Conclusiones
 - Basado en el análisis de datos históricos de incidentes (Shark Attack Dataset)

1. Validación del Mercado

     Alta Tasa de Supervivencia: El análisis de la columna Fatal muestra una abrumadora mayoría de casos no letales (4.394 supervivientes).

    Comparativa de Actividad: Al filtrar por actividades, hemos demostrado que el buceo (diving, scuba) representa un porcentaje manejable de incidentes en comparación con actividades de superficie como el surf, lo que nos da luz verde para operar bajo protocolos controlados.

2. Localizacion
    Tras analizar el "Top 3 Estados" (California, Florida, Hawaii), proponemos California como lugar donde construir nuestro negocio. 

    - California:

     Mayor cantidad de heridas físicas (45) y alta presencia de Tiburón Blanco (13 de 14 ataques en septiembre fueron de esta especie).

     La Estrategia: Debido a la especie involucrada (White Shark) y el tipo de riesgo, el buceo en jaula (Cage Diving) es obligatorio.

     Temporada Alta: Debemos operar a máxima capacidad en Julio, Agosto y Septiembre, que concentran los picos de avistamientos.

    - Florida:

     Alta incidencia de heridas físicas (41) y una tasa de fatalidad a tener en cuenta (mayor que CA en proporción a tus datos filtrados).

    - Hawaii:

     Menos incidentes físicos totales, pero una fatalidad proporcionalmente alta (igual a Florida con menos casos).


3. Perfil del Cliente 
    - Demografía: Al imputar la edad (mediana) y limpiar los datos, identificamos que el grueso de los incidentes ocurre en adolescentes y adultos jóvenes. 

    - Marketing: Nuestra publicidad debe de estar dirigida a un público joven (20-35 años) con una mediana de 24 años

4. Protocolos de Seguridad Basados en Datos 

   - California: Inversión prioritaria en Jaulas de Acero. El riesgo principal es "Herida Física".

   - Florida/Hawaii:Al tener casos de "Herida Física" y "Fatalidad" sin jaula, la velocidad de respuesta (torniquetes, hemostáticos) es lo que mantiene esa tasa de supervivencia alta.

- Basándonos en el análisis de datos, abrir una empresa de buceo en California especializada en avistamiento de Tiburón Blanco durante el verano (Jul-Sep) utilizando jaulas es la opción más rentable. Existe una    demanda clara (alta frecuencia de encuentros) y el riesgo es gestionable (baja letalidad relativa si se usa protección). Florida quedaria como segunda opción



In [75]:
porcentage_mes_injury = df_usa_diving['mes'].value_counts(normalize=True)*100
porcentage_mes_injury = round(porcentage_mes_injury,2)

In [76]:
df_calendario = porcentage_mes_injury.reset_index()
df_calendario.columns =['mes','probabilidad_de_encuentro']

df_calendario= df_calendario.sort_values(by='probabilidad_de_encuentro',ascending=False)
print(df_calendario.head(12))

    mes  probabilidad_de_encuentro
0   Sep                      15.74
1   Aug                      15.23
2   Jul                      13.71
3   Oct                      10.66
4   Jun                       8.63
5   May                       7.11
6   Nov                       7.11
7   Mar                       5.58
8   Dec                       4.57
9   Feb                       4.57
10  Jan                       4.06
11  Apr                       3.05


He calculado los porcentages del mes con mas probabilidades de encuentro, despues los he ordenado desde el mes con mas probabilidad al que menos

In [77]:
porcentage_riesgo = df_seguridad['Tipo_Riesgo'].value_counts(normalize=True)*100
porcentage_riesgo = round(porcentage_riesgo,2)

In [89]:
riesgo_california = df_seguridad[df_seguridad['state'] == 'California']['Tipo_Riesgo'].value_counts(normalize=True) * 100
round(riesgo_california,2)


Tipo_Riesgo
HERIDA FISICA    71.43
SIN HERIDAS      15.87
OTRO/LEVE         7.94
FATAL             4.76
Name: proportion, dtype: float64

Porcentages de los tipos de riesgos en California

In [92]:
riesgo_florida = df_seguridad[df_seguridad['state'] == 'Florida']['Tipo_Riesgo'].value_counts(normalize=True) * 100
round(riesgo_florida,2)

Tipo_Riesgo
HERIDA FISICA    56.94
OTRO/LEVE        16.67
SIN HERIDAS      15.28
FATAL            11.11
Name: proportion, dtype: float64

Porcentages de los tipos de riesgos en Florida

In [91]:
riesgo_hawaii = df_seguridad[df_seguridad['state'] == 'Hawaii']['Tipo_Riesgo'].value_counts(normalize=True) * 100
round(riesgo_hawaii,2)

Tipo_Riesgo
HERIDA FISICA    63.46
FATAL            15.38
SIN HERIDAS      11.54
OTRO/LEVE         9.62
Name: proportion, dtype: float64

Porcentages de los tipos de riesgos en Hawaii

In [97]:
porcetage_tabla_riesgos = pd.DataFrame({'Hawaii': riesgo_hawaii,'Florida': riesgo_florida,'Calfornia': riesgo_california})
redondeo_tabla_riesgos = round(porcetage_tabla_riesgos,2)
redondeo_tabla_riesgos

Hawaii  Florida  Calfornia
Tipo_Riesgo                              
FATAL           15.38    11.11       4.76
HERIDA FISICA   63.46    56.94      71.43
OTRO/LEVE        9.62    16.67       7.94
SIN HERIDAS     11.54    15.28      15.87

He creado un Data Frame con los porcentages de la tabla Fatal de cada estado. Podemos observar que donde menos Fatalidad hay es en California, con una diferencia de puntos porcentuales sobre Florida de mas de 6 puntos y mas de 11 puntos sobre Hawaii.
Debemos tener en cuenta que en califoria es donde mas heridas se sufren 71.43% pero tambien donde mas avistamientos sin heridas hay, 15.87%

In [108]:
print(df_usa['state'].unique())

['California' 'Sonoma County California' 'Hawaii West Coast Big Island'
 'Hawaii' 'Texas' 'Off California' 'Florida' 'Massachusetts'
 'Long Island NY' 'South Carolina' 'North Carolina' 'Floria' 'Galveston'
 'New York' ' Utah' 'New Jersey' 'Samoa' 'Louisiana' 'Mississippi'
 'Noirth Carolina' 'Georgia' 'Alabama' 'Los Angeles' 'Maryland' 'BAHAMAS'
 'Maui' 'Oregon' 'Franklin County, Florida' 'Virgin Islands' 'Maine'
 'Bahamas' 'Delaware' 'Guam' 'Cayman Islands' 'Rhode Island' 'Washington'
 nan 'Palmyra Atoll' 'Puerto Rico' 'Virginia' 'US Virgin Islands'
 'Kentucky' 'New Mexico' 'South Carolina ' 'Johnston Atoll' 'Alaska'
 'Missouri' 'North Carolina ' 'North & South Carolina' 'Carolina coast'
 'Florida ' 'Connecticut' 'Pennsylvania' 'Illinois' 'Wake Island'
 ' North Carolina' 'Midway Atoll' 'East coast' ' New Jersey' 'New York '
 'CUBA']


In [122]:
df_california = df_usa[df_usa['state'] == 'California']
df_california.value_counts()

date                  year  type        country  state       location                                                       activity                       name                           sex  injury                                        fatal    time           species      source                                                                                                           age_clean
2024-06-02 00:00:00   2024  Unprovoked  USA      California  Del Mar Beach San Diego                                        Swimming ocean training        Caleb Adams                    M    Significant bites to torso left arm and hand  N        9h             Unknown      ABC News San Diego: Del Mar Times:                                                                               46           1
01 Oct-2023           2023  Unprovoked  USA      California  Wildcat Beach, Marin County                                    Swimming                       Felix Louis N'Jai              M    Body not r

Filtramos por el estado de california

In [142]:
top_california = df_california['location'].value_counts(normalize=True).head(50)*100

round(top_california,2)


location
Huntington Beach, Orange County                                         2.15
San Onofre, San Diego County                                            1.53
La Jolla, San Diego County                                              1.53
Moonstone Beach, Humboldt County                                        1.23
Farallon Islands                                                        1.23
San Onofre State Beach, San Diego County                                1.23
Dillon Beach, Marin County                                              1.23
Franklin Point, San Mateo County                                        1.23
Santa Monica, Los Angeles County                                        1.23
Tomales Point, Marin County                                             1.23
Venice Beach, Los Angeles County                                        1.23
Malibu, Los Angeles County                                              0.92
Santa Barbara County                                               

He buscado el top de ciudades que aparecen en california, aparece mucho Los Angeles y Orange County 

In [146]:
len(df_california)



326

El total de localizaciones en california es 326

In [154]:
top_losangeles = r'Los Angeles| Orange County'
df_location_angeles = df_california[df_california['location'].astype(str).str.contains(top_losangeles, case=False, na=False)]
df_location_angeles

date  year        type country       state  \
380           30-Jun-2021  2021  Unprovoked     USA  California   
570           05-Oct-2019  2019  Watercraft     USA  California   
697           24-Aug-2018  2018  Unprovoked     USA  California   
873           03-May-2017  2017     Invalid     USA  California   
979           15-Jul-2016  2016  Unprovoked     USA  California   
...                   ...   ...         ...     ...         ...   
5930          04-Jul-1924  1924  Watercraft     USA  California   
6181          05-Sep-1908  1908  Unprovoked     USA  California   
6222             Sep-1906  1906    Provoked     USA  California   
6585           3-Jul-1879  1879  Unprovoked     USA  California   
7007  Before  19-Jun-1959     0  Unprovoked     USA  California   

                                         location  \
380                Catalina Island, Orange County   
570   Ship Rock Near Catalina, Los Angeles County   
697               Huntington Beach, Orange County   
873                   Sunset Beach, Orange County   
979                       Surfside, Orange County   
...                                           ...   
5930                 Newport Beach, Orange County   
6181                  Redondo, Los Angeles County   
6222                   Avalon, Los Angeles County   
6585              Los Angeles, Los Angeles County   
7007                    Capistrano, Orange County   

                                          activity  \
380                                       Canoeing   
570                                       Kayaking   
697                                       Swimming   
873                                        Surfing   
979                                   Kite surfing   
...                                            ...   
5930                                       Fishing   
6181  On fishing boat & trailing hand in the water   
6222                                       Fishing   
6585                                       Bathing   
7007                                           NaN   

                                                   name sex  \
380                                        Eddie Cahill   M   
570                                      Danny McDaniel   M   
697                                           Anonymous   M   
873                                         Sophia Raab   F   
979                                           Lee Frees   M   
...                                                 ...  ..   
5930  18' boat, occupants Richard Gunther & Donald C...   M   
6181                                       Ralph Nelson   M   
6222                                        Percy Neale   M   
6585                                           John Fry   M   
7007                                               girl   F   

                                                 injury    fatal   time  \
380                                        Hand injured        N  07h15   
570                          No injury, shark bit kayak        N  16h30   
697                   Shark bumped chest but no injury         N  20h00   
873   Laceration to thigh, likely caused by surfboar...  UNKNONW  14h30   
979                            No injury, board damaged        N  17h00   
...                                                 ...      ...    ...   
5930  No injury, shark tore hole in the side of the ...        N    NaN   
6181                             Lacerations to fingers        N    NaN   
6222  PROVOKED INCIDENT        No injury; shirt torn...        N    NaN   
6585                                         No details        N    NaN   
7007                                        Leg injured        N    NaN   

          species                                             source  \
380       Unknown  R. Collier, GSAF & K. McMurray, TrackingSharks...   
570   White Shark                 Kevin McMurray, TrackingSharks.com   
697   Other Shark                   K. McMurray, Trackin

In [155]:
len(df_location_angeles)

64

San Diego aparce 45 veces de 326

In [156]:
total_california = len(df_california)
total_angeles = len(df_location_angeles)
porcentage_losangeles = (total_angeles / total_california) * 100
round(porcentage_losangeles,2)

19.63

El porcentage de estas ubicaciones es de 19.63, las he puesto las dos juntas ya que son zonas vecinas
